use parallel ZIP methods for constructing JAR files #621

ricardoffmartins opened the pull request:
use parallel ZIP methods for constructing JAR files #621 …
e78ef05
pack methods created. (zipper.scala)

unpack methods created. (zipper.scala)

ref.: commons-compress

ricardoffmartins commented:
I changed the Makefile so i could test this feature.
Whenever i try to install my changes, it always restore the version 0.7.3, and it would remove the commons-compress binary.

odisseus commented:
Please rebase this pull request on top of the latest master, so that it gets checked by the new CI configuration.

ricardoffmartins commented:
@odisseus 
I having some trouble installing the last version of fury.

It look like it complains about this "INIT_CGROUP=$(shell cat /proc/1/cgroup | tail -n1 | cut -d: -f3)"

These folders "/procs/..." are from Unix-like systems, am i right?

odisseus commented:
@propensive, it looks like PR #695 causes problems on some systems that don't have cgroups, but are capable of running Fury.

ricardoffmartins commented:
@propensive 
@odisseus 
I changed the makefile so it doesnt look for cgroups, just to install, and i am still with problems. it looks related to env variables definitions 

`fury git:(master) ✗ make install 
mkdir -p dist/bundle/lib
mkdir -p opt
curl -C - -s -o opt/fury-0.7.11.sh "https://storage.googleapis.com/revivalist/downloads/fury.build/fury-0.7.11.sh"
sh opt/fury-0.7.11.sh opt/fury-0.7.11
     _____
    / ___/__ __ ____ __ __
   / __/ / // // ._// // /
  /_/    \_._//_/   \_. /
                   \___/

Fury build tool for Scala, version 0.7.11.
This software is provided under the Apache 2.0 License.
Copyright 2019 Jon Pretty, Propensive OÜ.

This will install Fury on your computer.

 * Installing Fury to opt/fury-0.7.11
 * Found Java version 1.8.0_171.
 * Installing Fury version 0.7.11...
 * Installed Fury 0.7.11.
 * Not updating shell configurations for batch-mode installation
 * Checking for currently-active Fury daemon...
touch opt/fury-0.7.11/bin/fury
mkdir -p bootstrap/bin
tar -cvzf bootstrap/build.fury .focus.fury layers/*
a .focus.fury
a layers/0A24A1586682A39838622B66BACF81A3B51A7F3448371AB31B54B75DA12DCD28
a layers/29B89F711E5D63E22B8D798F55C6FF01A4E1638C12C4F1DD631DEE4CBC48EF2B
a layers/85D48E3D72DC775FE785EA3B2520DAD39A79C3D76D2540C864D47329BE02897B
opt/fury-0.7.11/bin/fury layer extract -f bootstrap/build.fury
**opt/fury-0.7.11/bin/fury: line 6: realpath: command not found
cat: /classpath: No such file or directory**
Starting Fury daemon.....Error: Could not find or load main class com.facebook.nailgun.NGServer
....................................^CStarting Fury daemon.....Error: Could not find or load main class com.facebook.nailgun.NGServer
..........................................................................................................................................................................................`

odisseus commented:
Looks like you are missing the [`realpath`](https://linux.die.net/man/1/realpath) command, which is [used in the launcher script](https://github.com/propensive/fury/pull/681/files#diff-8798d09f69843a157d17afadbc3ad20cR6) since PR #681.

As a workaround, you can try to restore that line to the state from before that PR.

ricardoffmartins commented:
hese three simple steps are going to solve this and many other OS X issues:

Install Homebrew
brew install coreutils
grealpath .


Found this on Stackoverflow. Seems to fix the issue

odisseus commented:
@ricardoffmartins, both problems should now be fixed in the latest master. Please update your branch.

ricardoffmartins commented:
I removed my previous installation of coreutils, and it worked fine with the last changes.
 thanks you

ricardoffmartins commented:
@odisseus I rebased the branch. 

The CI fails due to this
`Could not find the file /home/runner/.local/share/fury/layers/9869BCB58C5888F63E6E46959BEA1D07A6F2DD6EC0CAE84DA23C37DA837EEDBB. Run `fury layer init` to create a new layer.`

I think this is fury layer with commons-compress binary that i added.  Because it didn't had it before. am i correct?

odisseus commented:
You have to update the layer files that are stored in the repository. Here's an instruction how to do it: https://github.com/propensive/fury/pull/691#issuecomment-555622258

ricardoffmartins commented:
@odisseus 
Is there a way to see the logs? 

I am getting
`Waiting 50ms until the bsp connection is up...
Waiting 50ms until the bsp connection is up...
Waiting 50ms until the bsp connection is up...
An unexpected error occurred which has been logged to /home/runner/work/fury/fury/.fury/logs/2019-11-26-17:57:50.916-7ece1192-0606-4739-b381-3d4a61112553.log
Waiting 50ms until the bsp connection is up...
make: *** [dist/bundle/lib/fury-frontend.jar] Error 1
Makefile:82: recipe for target 'dist/bundle/lib/fury-frontend.jar' failed
##[error]Process completed with exit code 2.`

but i can't check that log.

odisseus commented:
Please rebase again. The compilation logs can now be downloaded in case of a compilation failure.

However, the master branch cannot be compiled at the moment. If you see a `TimeoutException` in the log, this is caused by the broken master.

propensive commented:
I tried this out. It seems to work for the simple example I attempted, and it's about twice as fast. :)

Is it ready to merge yet?

odisseus commented on src/core/shell.scala:
```diff
@@ -161,11 +163,21 @@ case class Shell(environment: Environment) {
+    val result = new File(dest.value)
+    val zos = new ZipArchiveOutputStream(result)
+    zos.setEncoding("UTF-8")
+  /*  import org.apache.commons.compress.archivers.zip.ZipArchiveEntry
```
Uncomment or remove before merging

odisseus commented:
Oops! I'm afraid you'll have to re-create the layer files once again...

propensive commented:
@odisseus Don't worry, I'm just tidying up the PR before I merge it. I'll update the layers when I rebase.
